website/docs/configuration/types.html.md

```diff
@@ -0,0 +1,187 @@
+---
+layout: "docs"
+page_title: "Type Constraints - Configuration Language"
+sidebar_current: "docs-config-types"
+description: |-
+  Terraform module authors and provider developers can use detailed type
+  constraints to validate the inputs of their modules and resources.
+---
+
+# Type Constraints
+
+Terraform module authors and provider developers can use detailed type
+constraints to validate user-provided values for their input variables and
+resource arguments. This requires some additional knowledge about Terraform's
+type system, but allows you to build a more resilient user interface for your
+modules and resources.
+
+## Type Keywords and Construction Functions
+
+Type constraints are expressed using a mixture of _type keywords_ and
+function-like constructs called _type construction functions._
+
+* Type keywords are unquoted symbols that represent a static type.
+* Type constructor functions are unquoted symbols followed by a pair of
+  parentheses, which contain an argument that specifies more information about
+  the type. Without its argument, a type constructor function does not fully
+  represent a type; instead, it represents a _kind_ of similar types.
+
+Type constraints look like other kinds of Terraform
+[expressions](./expressions.html), but are a special syntax. Within the
+Terraform language, they are only valid in the `type` argument of an [input
+variable](./variables.html).
+
+## Primitive Types
+
+A _primitive_ type is a simple type that isn't made from any other types. All
+primitive types in Terraform are represented by a type keyword. The available
+primitive types are:
+
+* `string`: a sequence of Unicode characters representing some text, such
+  as `"hello"`.
+* `number`: a numeric value. The `number` type can represent both whole
+  numbers like `15` and fractional values such as `6.283185`.
+* `bool`: either `true` or `false`. `bool` values can be used in conditional
+  logic.
+
+### Conversion of Primitive Types
+
+The Terraform language will automatically convert `number` and `bool` values
+to `string` values when needed, and vice-versa as long as the string contains
+a valid representation of a number or boolean value.
+
+* `true` converts to `"true"`, and vice-versa
+* `false` converts to `"false"`, and vice-versa
+* `15` converts to `"15"`, and vice-versa
+
+## The "Any" Type
+
+The type keyword `any` is a special type constraint that accepts any value.
+
+## Complex Types
+
+A _complex_ type is a type that groups multiple values into a single value.
```

Comment:
I grouped these together into a super-category, since I saw the term "complex type" being used elsewhere in this section.